### ISSUE TITLE
Parser incorrectly evaluates "parserOptions.project" attribute in nested .eslitrc files

### ISSUE THREAD

**Issue by mtowers** [typescript-eslint/typescript-eslint#2320]
## Summary
When an `.eslintrc` file is nested in a folder in a project, the `parserOptions.project` attribute in the nested folder is evaluated relative to the _parent_ folder.  But the `extends` attribute is correctly evaluated relative to the current folder.

## Repro Steps

1. Create a TypeScript project with application code and test code separated into different directories at the project root.
1. Add a `.eslintrc.js` file to the project root
1. Add a `.eslintrc.js` file to the `./test` folder that extends the file in the project root.
1. Add a `tsconfig.json` file to the project root that `include`s only the application source in the `./src` folder.
1. Add a `tsconfig.json` file to the `./test` folder that `include`s only the test source in the `./test`folder.

**Expected Results**
- Source files in the `./test` folder should be linted as per the rules described in `./test/.eslintrc.js`.
- Source files in the `./src` folder should be linted as per the rules described in `./.eslintrc.js`.

**Actual Results**
Source files in the `./test` folder are not linted - linter reports an error of:
```
Parsing error: "parserOptions.project" has been set for @typescript-eslint/parser.
The file does not match your project config: test/test.spec.ts.
The file must be included in at least one of the projects provided.eslint
```

**Project directory structure**
```
eslint-test/
  |-src/
    |-app.ts
  |-test/
    |-.eslintrc.js
    |-test.spec.ts
    |-tsconfig.json
  |-.eslintrc.js
  |-tsconfig.json
  |-tsconfig.test.json
```

**`eslint-test/test/.eslintrc.js`**
```JS
module.exports = {
    extends: "../.eslintrc.js",  // resolves to file in parent directory (note leading '..')
    parserOptions: {
        project: "./tsconfig.json" // resolves to file in *parent* directory - causes error
        // "project": "../tsconfig.test.json"  // (note leading '..') attempts to find file in eslint-test/../
        // "project": "./tsconfig.test.json"  // lints test files correctly - see 'include' directive in file
    },
    rules: {
    }
}
```
**`eslint-test/test/test.spec.ts`**
```TS
/**
 * Test code
 */
console.log('hello test')
```

**`eslint-test/test/tsconfig.json`**
```JSON
{
    "extends": "../tsconfig.json",
    "compilerOptions": {
        "experimentalDecorators": true,
        "noUnusedLocals": false
    },
    "include": ["./**/*.ts"]
}
```
**`eslint-test/.eslintrc.js`**
```JS
module.exports = {
    env: {
      browser: true,
      es6: true
    },
    'extends': [
      'plugin:@typescript-eslint/all',
      'prettier',
      'prettier/@typescript-eslint'
    ],
    ignorePatterns: [
      '.eslintrc.js'
    ],
    parser: '@typescript-eslint/parser',
    parserOptions: {
      project: './tsconfig.json',
      sourceType: 'module'
    },
    plugins: [
      '@typescript-eslint',
      '@typescript-eslint/tslint'
    ],
    rules: {
      // rules collapsed for brevity
    }
}
```
**`eslint-test/tsconfig.json`**
```JS
{
    "compilerOptions": {
       // options collapsed for brevity
    },
    "exclude":[
        "node_modules"
    ],
    "include": ["src/**/*.ts"]
}
```
**`eslint-test/tsconfig.test.json`**
```JS
{
    "extends": "./tsconfig.json",
    "compilerOptions": {
        "experimentalDecorators": true,
        "noUnusedLocals": false
    },
    "include": ["./test/**/*.ts"]
}
```
## Additional Info

**eslint error output** (from VSCode Output console)
```
eslint:linter Parsing error: "parserOptions.project" has been set for @typescript-eslint/parser.
The file does not match your project config: test/test.spec.ts.
The file must be included in at least one of the projects provided.
Error: "parserOptions.project" has been set for @typescript-eslint/parser.
The file does not match your project config: test/test.spec.ts.
The file must be included in at least one of the projects provided.
```
Related Tickets:
- https://github.com/typescript-eslint/typescript-eslint/issues/967
- https://github.com/typescript-eslint/typescript-eslint/issues/1531
- https://github.com/typescript-eslint/typescript-eslint/issues/1723

## Versions

| package                     | version |
| --------------------------- | ------- |
| `@typescript-eslint/parser` | `2.21.0` |
| `TypeScript`                | `3.9.7` |
| `ESLint`                    | `6.8.0` |
| `node`                      | `12.18.2` |
| `yarn`                       | `1.22.4` |

**Comment by bradzacher** [comment]
ESLint doesn't tell us anything about the paths of config files or anything of the sort.
This is done on purpose because you can specify configs in many different ways, some of which have no folder (i.e. CLI flags).

This is why we have the `tsconfigRootDir` parser option:
https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/parser#parseroptionstsconfigrootdir

**Comment by mtowers** [comment]
@bradzacher - Thanks large for the clarification.  Does `tsconfigRootDir` need to be an absolute path?
In my `eslint-test/.eslintrc.js` file I set `tsconfigRootDir` to `./` and got the same behavior as described above.

**Comment by bradzacher** [comment]
as mentioned - eslint doesn't pass us any context about the config files - we calculate _everything_ relative to the CWD.

So setting it to `./` leaves you in exactly the same position as before.
We recommend setting it to `__dirname` for simplicity

**Comment by mtowers** [comment]
## Solved!
For whomever stumbles upon this next, the solution is to set `tsconfigRootDir: __dirname` in the _nested_ `.eslintrc.js` file.  In the example above, `eslint-test/test/.eslintrc.js` should look like the following.  Note:  This requires that you use JavaScript `.js` files for your linter rules, not JSON `.json`.
```JS
module.exports = {
    extends: "../.eslintrc.js",  // resolves to file in parent directory
    parserOptions: {
        project: "tsconfig.json",
        tsconfigRootDir: __dirname
    },
    rules: {
    }
}
```

**Comment by bradzacher** [comment]
side note: `.json` config files are going to be deprecated and removed in a future version of ESLint, so you should switch to `.js` config files anyway.
